jena-arq/src/main/java/org/apache/jena/riot/other/G.java

```diff
@@ -0,0 +1,669 @@
+        Objects.requireNonNull(graph, "graph");
+        Objects.requireNonNull(type, "type");
+        List<Node> types = listSubClasses(graph, type);
+        List<Node> nodes = new ArrayList<>();
+        accNodesOfTypes(nodes, graph, types);
+        return nodes;
+    }
+
+    /**
+     * List all the types of a node, including super-classes.
+     */
+    public static Set<Node> allTypesOfNodeRDFS(Graph graph, Node node) {
+        Objects.requireNonNull(graph, "graph");
+        Objects.requireNonNull(node, "node");
+        Set<Node> types = typesOfNodeAsSet(graph, node);
+        Set<Node> types2 = new HashSet<>();
+        types.forEach(t->{
+            List<Node> subClasses = listSuperClasses(graph, t);
+            types2.addAll(subClasses);
+        });
+        return types2;
+    }
+
+    /** List all the node of type, including considering rdfs:subClassOf */
+    public static Set<Node> allNodesOfTypeRDFS(Graph graph, Node type) {
+        Objects.requireNonNull(graph, "graph");
+        Objects.requireNonNull(type, "type");
+        Set<Node> types = subClasses(graph, type);
+        Set<Node> nodes = new HashSet<>();
+        accNodesOfTypes(nodes, graph, types);
+        return nodes;
+    }
+
+    /** For each type, find nodes of that type and accumulate */
+    private static void accNodesOfTypes(Collection<Node> acc, Graph graph, Collection<Node> types) {
+        types.forEach(t->
+            find(graph, null, rdfType, t).mapWith(Triple::getSubject).forEach(acc::add)
+            );
+    }
+
+    /** Return a set of all objects for subject-predicate */
+    public static Set<Node> allSP(Graph graph, Node subject, Node predicate) {
+        Objects.requireNonNull(graph, "graph");
+        return find(graph, subject, predicate, null).mapWith(Triple::getObject).toSet();
+    }
+
+    /** Return a set of all subjects for predicate-object */
+    public static Set<Node> allPO(Graph graph, Node predicate, Node object) {
+        Objects.requireNonNull(graph, "graph");
+        return find(graph, null, predicate, object).mapWith(Triple::getSubject).toSet();
+    }
+
+    // --- Graph walking.
+
+    /** Count the number of in-arc to an object */
+    public static long objectConnectiveness(Graph graph, Node object) {
+        Objects.requireNonNull(graph, "graph");
+        Objects.requireNonNull(object, "object");
+        return Iter.count(find(graph, null, null, object));
+    }
+
+    /** Test whether an object has exactly one in-arc. */
+    public static boolean oneConnected(Graph graph, Node object) {
+        Objects.requireNonNull(graph, "graph");
+        Objects.requireNonNull(object, "object");
+        ExtendedIterator<Triple> iter = find(graph, null, null, object);
+        if ( ! iter.hasNext() )
+            // Zero.
+            return false;
+        iter.next();
+        if ( iter.hasNext() )
+            // more than one
+            return false;
+        return true;
```

Comment:
Do we have to close `iter`?

Reply:
Yes. (It is also most the same as findUnique)

Closing unfinished extended iterators is contract even if many (most?) apps don't care or attend to the detail. In fact, plain in-memory, TDB1, TDB2 and TIM don't rely on this. But as a library it ought to get it right!